orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,156 @@ class PipelineSplitter {
+      auto& info = plan[batch];
+      info.events.resize(2 * num_stages - 1);  // two pairs for each stage except for last stage, which only need one
+      info.retired_batch = -1;
+    }
+
+    // each stage requires 2 pair of wait/record events for FW/BW
+    // except for last stage, which only need one pair of wait/record
+    for (size_t s = 0; s < num_stages; ++s) {
+      int64_t prev_event_id = -1;
+      int64_t event_id = start_ids[s];
+      int64_t retired_batch = -1;
+      for (size_t t = 0; t < timeline.GetNumSlots(); ++t) {
+        const auto& slot = timeline.Get(s, t);
+        if (!slot.occupied)
+          continue;
+
+        ORT_ENFORCE(event_id < max_id_);
+        if (slot.is_backward) {
+          if (s < num_stages - 1) {
+            // dont update wait event for last stage's bw
+            plan[slot.batch_id].events[2 * num_stages - 2 - s].first = prev_event_id;
+            prev_event_id = event_id;
+          }
+          plan[slot.batch_id].events[2 * num_stages - 2 - s].second = event_id++;
+          retired_batch = gsl::narrow<int64_t>(slot.batch_id);
+        } else {
+          if (s == 0) {
+            // update retired batch when starting a new batch (s == 0 && !bw)
+            plan[slot.batch_id].retired_batch = retired_batch;
+          }
+          plan[slot.batch_id].events[s].first = prev_event_id;
+          prev_event_id = event_id;
+          if (s < num_stages - 1) {
+            // dont update record event for last stage's fw
+            plan[slot.batch_id].events[s].second = event_id++;
+          }
+        }
+      }
+    }
+  }
+
+ private:
+  class Timeline {
+   public:
+    struct Slot {
+      size_t batch_id;
+      bool is_backward;
+      bool occupied;
+
+      Slot() : occupied(false) {}
+    };
+
+    Timeline(size_t stages, size_t num_slots) : slots_(stages) {
+      for (size_t s = 0; s < stages; ++s) {
+        slots_[s].resize(num_slots);
+      }
+    }
+
+    bool IsOccupied(size_t s, size_t t) const {
+      return slots_[s][t].occupied;
+    }
+
+    const Slot& Get(size_t s, size_t t) const {
+      return slots_[s][t];
+    }
+
+    size_t GetNumSlots() const {
+      return slots_[0].size();
+    }
+
+    void Occupy(size_t s, size_t t, size_t batch_id, bool bw) {
+      Slot& slot = slots_[s][t];
+      slot.occupied = true;
+      slot.is_backward = bw;
+      slot.batch_id = batch_id;
+    }
+
+   private:
+    std::vector<std::vector<Slot>> slots_;
+  };
+
+  Timeline GenerateTimeline(size_t num_stages, size_t num_batches) {
+    // 2 slots for FW/BW in each batch, plus 2 * (num_stages - 1) gaps
+    size_t num_slots = num_batches * 2 + 2 * (num_stages - 1);
+    Timeline timeline(num_stages, num_slots);
+    std::vector<size_t> t(num_stages);
+    std::vector<size_t> t_bw(num_stages);
+    for (size_t s = 0; s < num_stages; ++s) {
+      t[s] = 0;
+      t_bw[s] = 0;
+    }
+
+    // generate timeline
+    for (size_t batch = 0; batch < num_batches; ++batch) {
```

Comment:
>for (size_t batch = 0; batch < num_batches; ++batch) { [](start = 4, length = 54)

a brief description on the planning logic here: 
always prioritize bw to minimize the life span of a batch